### PR TITLE
add keep alive support to http/https connections

### PR DESCRIPTION
**Pull request by tsybinAndrey**
this agent options make http/https connections reusable and not creating new connection on every client request/.

### CODE REVIEW

**Comment by SimenB** [comment]
Agreed we shouldn't change the default. Allowing people to pass `agent` if that's not already supported seems like a cleaner option that's not as opinionated

**Comment by zbjornson** [comment]
Oof, right, isn't this already possible by providing the `agent` in the request options parameter?

**Comment by tsybinAndrey** [comment]
> Oof, right, isn't this already possible by providing the `agent` in the request options parameter?

Yep, this is possible and we what we've done right now is pass custom agent to http, but in my opinion this is not what is expected. For instance, almost all http agent libraries have keepAlive option turned on by default. It is obvious and convenient to create and reuse one connection per multiple requests when you send requests to pushgateway periodically.

File: CHANGELOG.md [hunk 1]
```diff
@@ -41,6 +41,7 @@ project adheres to [Semantic Versioning](http://semver.org/).
 - chore: replace `utils.getPropertiesFromObj` with `Object.values`
 - chore: remove unused `catch` bindings
 - chore: upgrade Prettier to 2.x
+- chore: add keep-alive support to client requests in pushgateway
```
[the remark this answers is not in the document]
Comment (tsybinAndrey) [reply CHANGELOG.md]: Moved

File: lib/pushgateway.js [hunk 2]
```diff
@@ -5,6 +5,16 @@ const http = require('http');
 const https = require('https');
 const { globalRegistry } = require('./registry');
 
+const keepAliveConfig = {
+	keepAlive: true,
+	keepAliveMsecs: 10000,
+	maxSockets: 5,
+};
+
+const httpAgent = new https.Agent(keepAliveConfig);
```
[the remark this answers is not in the document]
Comment (tsybinAndrey) [reply lib/pushgateway.js]: Yep, changed it to http

**Comment by tsybinAndrey** [comment]
> Thanks for opening this!
> 
> I'm +1 on adding this, but I think it would be good to preserve the default behavior (use the global Agent). I could be persuaded though that this is a better default. @SimenB @siimon ? If we do change the default, it's semver-major, and the changelog should have more info about this.
> 
> I'd also document how to configure these options in the README (i.e. show how to provide your own Agent to the gateway).

I've made some fixes and also add example to readme on how to change default agent configuration.

**Comment by SimenB** [comment]
>For instance, almost all http agent libraries have keepAlive option turned on by default.

It's not turned on by default in Node, so I don't think we should turn it on by default either. I'm very much down with documenting how to do it with an explicit example, though

**Comment by siimon** [comment]
I agree with not changing the default behaviour but make it configurable! 👍

**Comment by zbjornson** [comment]
Given that the consensus was to not change the default behavior, I just added some text to the readme showing how to configure the agent.

**Pull request closed without merging**